gdal/ogr/ogrsf_frmts/cad/ogrcadlayer.cpp

```diff
@@ -0,0 +1,639 @@
+/*******************************************************************************
+ *  Project: OGR CAD Driver
+ *  Purpose: Implements driver based on libopencad
+ *  Author: Alexandr Borzykh, mush3d at gmail.com
+ *  Language: C++
+ *******************************************************************************
+ *  The MIT License (MIT)
+ *
+ *  Copyright (c) 2016 Alexandr Borzykh
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining a copy
+ *  of this software and associated documentation files (the "Software"), to deal
+ *  in the Software without restriction, including without limitation the rights
+ *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *  copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in all
+ *  copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *******************************************************************************/
+#include "ogr_cad.h"
+#include "cpl_conv.h"
+
+#include <sstream>
+#include <iomanip>
+
+OGRCADLayer::OGRCADLayer( CADLayer &poCADLayer_, OGRSpatialReference *poSR ) :
+    poCADLayer( poCADLayer_ )
+{
+    nNextFID = 0;
+    poSpatialRef = NULL;
+
+    poFeatureDefn = new OGRFeatureDefn( CADRecode( poCADLayer_.getName(), 29 ) );
+
+    // Setting up layer geometry type
+    OGRwkbGeometryType eGeomType;
+    switch( poCADLayer.getGeometryType() )
+    {
+        case CADObject::ATTDEF:
+        case CADObject::TEXT:
+        case CADObject::MTEXT:
+        case CADObject::POINT:
+            eGeomType = wkbPoint;
+            break;
+        case CADObject::ELLIPSE:
+        case CADObject::ARC:
+        case CADObject::CIRCLE:
+        case CADObject::POLYLINE3D:
+        case CADObject::POLYLINE2D:
+        case CADObject::LWPOLYLINE:
+        case CADObject::LINE:
+            eGeomType = wkbLineString;
+            break;
+        case CADObject::FACE3D:
+            eGeomType = wkbPolygon;
+            break;
+        case -1:
```

Comment:
The thing is that CADObject enum is used in library to represent cross dxf/dwg object type. By this idea, I think its not a real case when I should add value CADObject::GeometryBag (?) because it's not a CAD geometry at all. poCADLayer.getGeometryType() returns a short int, so -1 was chosen as a constant for a 'mixed geometry type' layer.